### DISCUSSION TITLE
Which best practices for lua are applicable to neovim's luajit and should there be a resource for such recommendations?

### DISCUSSION THREAD

**Discussion by sandersantema** [neovim/neovim#24327] [General]
There are some resources online about best practices in regards to lua performance, such as this [lua gems chapter](https://www.lua.org/gems/sample.pdf). But what isn't clear to me is whether this is applicable to luajit and specifically to luajit as it is running under neovim.

Some examples of these best practices for lua are
- using `string.format` instead of string concatenation (`..`)
- declaring local functions instead of using closures
- using local variables, e.g. calling `string.format` is more expensive than calling `format` when defined as a local `local format = string.format`

Looking through some plugins I see that these best practices are often not followed. I wanted to improve this and open a pr for one such plugin which is performance critical but I'm not at all certain whether these best practices hold for luajit, hence this discussion.
Furthermore taking a cursory look at the neovim itself (`rg '[^\.]\.\.[^\.]' --type lua -g '!*_spec.lua' -g '!*test*'` that is), it seems `..` is used quite a lot too.

To be clear: I'm not referring to best practices in regards to plugin development and such but rather to best practices for performance specific to luajit and neovim.

I think it would be a good idea if there were some general recommendations of best practices in regards to performance, which could hopefully be based on measurements and expertise and because of that hopefully not that controversial either.

**Comment by clason** [comment]
Most performance recommendations for LuaJIT are specifically for numerical code, which is unlikely to be relevant to Neovim plugins (or config):
https://kress95.github.io/luajit-wiki/Numerical-Computing-Performance-Guide/

So the basic recommendations are the same. In particular: always benchmark before you try to optimize. If something is not a bottleneck for your code, readability is (much!) more important than shaving a few percent off. (All the things you mention tend to be negligible compared to the _actual_ workload you have to perform.)
